### PR TITLE
jest updated - corrected Matchers, added ExtendedExpect

### PR DESCRIPTION
**Pull request by tonyhallett**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: 

for change to message type
[getMessage](https://github.com/facebook/jest/blob/cc74b5159cbe2eb61eaa4d96b37a91eef1bef4c9/packages/expect/src/index.ts#L139)
and [usage](https://github.com/facebook/jest/blob/cc74b5159cbe2eb61eaa4d96b37a91eef1bef4c9/packages/expect/src/index.ts#L269)

for CustomExpect function [custom matchers added here](https://github.com/facebook/jest/blob/cc74b5159cbe2eb61eaa4d96b37a91eef1bef4c9/packages/expect/src/index.ts#L102) and [custom asymmetric matchers added here](https://github.com/facebook/jest/blob/cc74b5159cbe2eb61eaa4d96b37a91eef1bef4c9/packages/expect/src/jestMatchersObject.ts#L84)

For changes to Matchers<T> [see here](https://github.com/facebook/jest/blob/cc74b5159cbe2eb61eaa4d96b37a91eef1bef4c9/packages/expect/src/index.ts#L102) - there is no such thing as expect('').not.not.not....... that the [current typing allows](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/4f0f69b54a4b0a9f766054b1f92199fb695ee925/types/jest/index.d.ts#L618)
```
    interface Matchers<R> {
        
        /**
         * If you know how to test something, `.not` lets you test its opposite.
         */
        not: Matchers<R>;
```

Also [see here](https://github.com/facebook/jest/blob/cc74b5159cbe2eb61eaa4d96b37a91eef1bef4c9/packages/expect/src/index.ts#L239) for signature expect('').builtinMatcher('').  Similar issue existed for the promises.
The current definition file has
```
<T = any>(actual: T): Matchers<T>;
interface Matchers<R> {
lastCalledWith(...args: any[]): R;
```
R should be void and should be `Promise<void>` for .resolves/rejects

- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

I have not updated the built in asymmetric matchers as they return any...
Also the this context for matchers looks dated although changes may have been introduced since jest 24.0 that the definition file is supporting.

Finally in jest-tests.ts there are some tests that demonstrate the exactness of the new definition file over the previous and can probably be removed.
```
/* Corrections of previous typings */
// $ExpectError
expect('').not.not;
// $ExpectError
expect('').resolves.resolves;
// $ExpectType void
expect('').toEqual('');
// $ExpectType Promise<void>
expect(Promise.resolve('')).resolves.toEqual('');

// $ExpectError
const customMatcherResultMessage: jest.CustomMatcherResult['message'] = 'msg';
```

### CODE REVIEW

**Comment by tonyhallett** [comment]
I see that the declaration merging is causing issues...

**Comment by tonyhallett** [comment]
I will slightly change the jest definition and update the 7 repos that depend on it to use the new `interface Matchers<R, T> ` instead of the old `interface Matchers<R> {`.

**Comment by tonyhallett** [comment]
@mike-d-davydov @JoshuaKGoldberg @tkrotoff @dawnmist @deadNightTiger @lvl99 @joemitchard @mistic100 

Hi.  I have made some changes to jest definitions that required changes to yours - for most it is an additional generic parameter otherwise a type name change.

please review this PR in the next few days. Be sure to explicitly select Approve or Request Changes in the GitHub UI so I know what's going on.

If no reviewer appears after a week, a DefinitelyTyped maintainer will review the PR instead.

**Review by mistic100 — APPROVED** [review]

**Review by favna — APPROVED** [review]

**Comment by tonyhallett** [comment]
Hi.  I have a few changes that will be completed tomorrow.
Some additional - expect.getState/setState/matcher this context.
Others - simplify the new typing I added regarding Matchers<T> and some helper mapping types related to this.

Would it be preferable to close this pull request or push further commits ?

**Review by ExE-Boss — COMMENTED** [review]
The intersection operator&nbsp;`&` should&nbsp;be&nbsp;surrounded by&nbsp;spaces, like&nbsp;what&nbsp;[**Prettier**](https://prettier.io)&nbsp;does.

**Comment by tonyhallett** [comment]
@ExE-Boss I thought that would be caught by dtslint.  Have instead created `type MatcherContext = MatcherUtils & Readonly<MatcherState>;`

File: types/jest/jest-tests.ts [hunk 1]
```diff
@@ -696,7 +718,7 @@ expect.extend({
 });
 
 expect.extend({
-    foo(this: jest.MatcherUtils) {
+    foo(this: jest.MatcherUtils&Readonly<jest.MatcherState>) {
```
[the remark this answers is not in the document]
Comment (tonyhallett) [reply types/jest/jest-tests.ts]: Thanks, resolved in new commit

File: types/jest/index.d.ts [hunk 2]
```diff
@@ -427,14 +428,15 @@ declare namespace jest {
...
+        [other: string]: any;
     }
 
     interface ExpectExtendMap {
         [key: string]: CustomMatcher;
     }
 
     type CustomMatcher = (
-        this: MatcherUtils,
+        this: MatcherUtils&Readonly<MatcherState>,
```
[the remark this answers is not in the document]
Comment (tonyhallett) [reply types/jest/index.d.ts]: Thanks, resolved in new commit

**Comment by yoursunny** [comment]
This is a breaking change for any TypeScript project that uses `expect.extend`:
https://travis-ci.com/yoursunny/NDNts/builds/133646120

There's no way back now. However, at the very least, the advice at
https://github.com/facebook/jest/blob/f7eb09b3df76121df1c2499df1ba7d17242a0a8d/docs/ExpectAPI.md#L68
should be amended to say `Matchers<R, T>` instead of `Matchers<R>`.

**Comment by davidje13** [comment]
Can you give some clarity on what the new `T` parameter is for `Matchers`? As noted by yoursunny, this has broken any project which uses custom matchers and I'd like to understand what I'm supposed to use this new parameter for. In this code it seems to be used only by the snapshot matchers, and I'm not clear on how it differs from `R` in that scenario.

**Comment by tonyhallett** [comment]
@davidje13 The type T is as before, `expect<T = any>(actual: T): .....`.  The additional type parameter T on `Matchers<R, T>` is so that custom matchers can use T like toMatchSnaphot and for the return value to be typed correctly, which it was not before.  R should either be void or Promise<void> ( resolves/rejects ) and the new typing corrects this and other issues with the Matchers interface.

**Comment by davidje13** [comment]
Got it. I actually thought that's what `R` was for before but looking again I see that it's the return value (clearly I hadn't looked very closely before!)

**Pull request merged**